### PR TITLE
test(crypto): complete unit tests for the crypto package

### LINKED ISSUE #928: Crypto package unit tests
Our test suite for the crypto package is currently lacking thorough testing of all of the exposed functions and classes. As developers we should aim to test our code properly to ensure it works as expected.

### PR DESCRIPTION
**Pull request by ctrlc03**
# Description

Add unit tests for functions within the crypto package and fix errors that arised by the test cases. Note that babyJubAddPoint seem to not work, however this is not used by the codebase

## Related issue(s)

fix #928 

## Confirmation

- [x] I have read and understand MACI's [contributor guidelines](../CONTRIBUTING.md) and [code of conduct requirements](../CODE_OF_CONDUCT.md).
- [x] I have read and understand MACI's [GitHub processes](https://github.com/privacy-scaling-explorations/maci/discussions/847) documentation.
- [x] I have read and understand MACI's [testing guide](https://maci.pse.dev/docs/testing).

### CODE REVIEW

**Comment by kittybest** [comment]
https://github.com/privacy-scaling-explorations/maci/blob/ab870e84f58fae02db1acec78f0310f0f8c2a2ab/crypto/ts/utils.ts#L52

Not sure should we modify it here, but I found the `depth` doesn't have a description.

**Comment by ctrlc03** [comment]
> https://github.com/privacy-scaling-explorations/maci/blob/ab870e84f58fae02db1acec78f0310f0f8c2a2ab/crypto/ts/utils.ts#L52
> 
> Not sure should we modify it here, but I found the `depth` doesn't have a description.

good catch, will add it here as it's just one comment :)

File: crypto/ts/__tests__/Crypto.test.ts [hunk 1]
```diff
@@ -8,133 +10,730 @@ import {
...
+      });
+      it("should produce a non zero value", () => {
+        const hash = hash2([BigInt(1), BigInt(2)]);
+        expect(hash).to.not.eq(BigInt(0));
+      });
+      it("should throw when elements is more than numElement", () => {
+        expect(() => hashN(2, [BigInt(1), BigInt(2), BigInt(3)])).to.throw(
+          "the length of the elements array should be at most 2; got 3",
+        );
+      });
```
[the remark this answers is not in the document]
Comment (ctrlc03) [reply crypto/ts/__tests__/Crypto.test.ts:259]: Good catch, I changed there and on others too

File: crypto/ts/__tests__/Crypto.test.ts [hunk 2]
```diff
@@ -8,133 +10,730 @@ import {
...
+      it("should produce a non zero value", () => {
+        const hash = hash3([BigInt(1), BigInt(2), BigInt(3)]);
+        expect(hash).to.not.eq(BigInt(0));
+      });
+      it("should throw when elements is more than numElement", () => {
+        expect(() => hashN(3, [BigInt(1), BigInt(2), BigInt(3), BigInt(4)])).to.throw(
+          "the length of the elements array should be at most 3; got 4",
+        );
+      });
```
[the remark this answers is not in the document]
Comment (ctrlc03) [reply crypto/ts/__tests__/Crypto.test.ts:284]: you are right, we should have tested hash3 there, amended that

File: crypto/ts/__tests__/Crypto.test.ts [hunk 3]
```diff
@@ -8,133 +10,730 @@ import {
...
-      const expectedLength = plaintext.length <= 3 ? 4 : 1 + (plaintext.length % 3) * 3;
+      it("should always generate the same commitment for the same inputs", () => {
+        const commitment = genTreeCommitment(leaves, salt, depth);
+        expect(commitment).to.satisfy((num: bigint) => num > 0);
+        expect(commitment).to.satisfy((num: bigint) => num < SNARK_FIELD_SIZE);
 
-      expect(ciphertext.length).to.eq(expectedLength);
+        const commitment2 = genTreeCommitment(leaves, salt, depth);
+        expect(commitment).to.satisfy((num: bigint) => num > 0);
+        expect(commitment).to.satisfy((num: bigint) => num < SNARK_FIELD_SIZE);
```
[the remark this answers is not in the document]
Comment (ctrlc03) [reply crypto/ts/__tests__/Crypto.test.ts:515]: good catch

File: crypto/ts/__tests__/Crypto.test.ts [hunk 4]
```diff
@@ -8,133 +10,730 @@ import {
...
+        }).to.throw();
+      });
     });
   });
 
-  describe("Signature generation and verification", () => {
+  describe("signatures", () => {
+    const { privKey, pubKey } = genKeypair();
     const message = BigInt(Math.floor(Math.random() * 1000000000));
     const signature = sign(privKey, message);
```
Comment (kittybest) [inline crypto/ts/__tests__/Crypto.test.ts:739]: L740: typo --> constituent

File: crypto/ts/__tests__/Crypto.test.ts [hunk 5]
```diff
@@ -8,133 +10,730 @@ import {
...
+      it("should produce a result which is < SNARK_FIELD_SIZE", () => {
+        const pk = genPubKey(genPrivKey());
+        const pkBuff = packPubKey(pk);
+        const pkUnpacked = unpackPubKey(pkBuff);
+        expect(pkUnpacked[0] < SNARK_FIELD_SIZE).to.eq(true);
+        expect(pkUnpacked[1] < SNARK_FIELD_SIZE).to.eq(true);
+      });
+    });
+    describe("genPubKey", () => {
+      it("should produce a public key which is <= SNARK_FIELD_SIZE", () => {
```
[the remark this answers is not in the document]
Comment (ctrlc03) [reply crypto/ts/__tests__/Crypto.test.ts]: That's my mistake, we should be using `<` across all of the tests and code.

**Review by kittybest — APPROVED** [review]

File: crypto/ts/__tests__/Crypto.test.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (kittybest) [reply crypto/ts/__tests__/Crypto.test.ts]: that's consistent with my understanding, thanks a lot~~~

**Pull request merged**